rfcs/RFC-0002-configuration.md

```diff
@@ -0,0 +1,350 @@
+        {"role": "user", "content": "Write a haiku about coding"},
+    ],
+)
+```
+
+This script switches the ollama URL of an existing server to point to another ollama server. I used the ollama URL as the first field just because it was easy to test. The actual user configurable fields will at first be minimal just to get this API off the ground but can be easily expanded.  I also intend to expose this API through the llama-stack-client with the following commands:
+
+`llama-stack-client configurations inspect`
+`llama-stack-client configurations register –config-path –config…`
+
+## Some Key Features
+
+### Class Structure
+
+Here is what I have so far in terms of net-new API classes inside of llama-stack:
+
+
+```python
+@json_schema_type
+class Configuration(BaseModel):
+    type: Literal[ResourceType.configuration.value] = ResourceType.configuration.value
+    config: StackRunConfig
+
+
+@runtime_checkable
+@trace_protocol
+class Configurations(Protocol):
+    """Llama Stack Configuration API for storing and applying hyperparameters for given tasks."""
+
+    @webmethod(route="/configurations/register", method="POST")
+    async def register_config(
+        self,
+        config,
+    ) -> dict[str, Any]: ...
+```
+
+And for /v1/inspect:
+
+```python
+@runtime_checkable
+class Inspect(Protocol):
+
+    @webmethod(route="/inspect/configurations", method="GET")
+    async def inspect_config(
+        self,
+    ) -> InspectConfigResponse: ...
+```
+
+Where this is InspectConfigResponse:
+
+```python
+class InspectConfigResponse(BaseModel):
+    data: UserConfig
+```
+
+And this is a UserConfig: (There is a lot of overlap in methods, this is all hacking code!)
+
+```python
+class UserConfig(BaseModel):
+    providers: Dict[str, List[Provider]] = Field(
+        description="""
+One or more providers to use for each API. The same provider_type (e.g., meta-reference)
+can be instantiated multiple times (with different configs) if necessary.
+""",
+    )
+
+    @classmethod
+    def from_stack_run(
+        cls, registry: Dict[Any, Dict[str, Any]], stack_run: "StackRunConfig"
+    ) -> "UserConfig":
+        """
+        This is almost a method to go backwards and get a user config from an existing run config
+        """
+        user_config: Dict[str, List[Provider]] = {}
+        for type, providers in stack_run.providers.items():
+            api = Api(type)
+            user_config[type] = []
+            for provider in providers:
+                provider_config = {}
+                provider_spec = registry[api][provider.provider_type]
+                config_type = instantiate_class_type(provider_spec.config_class)
+                try:
+                    if provider.config:
+                        existing = config_type(**provider.config)
+                        for field_name, field in existing.model_fields.items():
+                            if field.json_schema_extra:
```

Comment:
we also need to check that `user_field` key in `field.json_schema_extra`  is `True`, right?